api/v1/scripts/localisation.py

```diff
@@ -0,0 +1,224 @@
+import json
+import hashlib
+import xml.etree.ElementTree as ET
+import os
+import logging
+import argparse
+
+STRING_KEYS = ('title', 'text', 'asQuestion', 'details', 'description')
+
+
+def build_string_index(path_questionnaire):
+	"""
+	Extract all strings from a questionnaire that should be translated.
+	"""
+	string_index = {}
+
+	def obj_hook_string_extraction(obj):
+		# for each key check if it is available
+		for key in STRING_KEYS:
+			if key in obj:
+				# calculate an ID for this string and save it to index
+				string = obj[key].strip()
+				string_hash = hashlib.md5(string.encode()).hexdigest()
```

Comment:
we should use the ID of the question for this. 
For the answers or options, we should use their respective Ids.

Reply:
No, I think md5 is more suitable, because this allows for reusing translations in several places. What would be the advantage of question/answer ids?

Reply:
human-readable and also unique. This also allows reusing because you can compare them between the questionnaires

Reply:
the md5 would change if you change to source string.

Reply:
Thats the point, because this invalidates the corresponding translations if the source changes. And I think this is a big advantage of MD5 which we would loose if we use question IDs.
On the other hand, we are heading for a graphical editor, so - best case - the user does not need to see any ID at all.
A common way to provide context in the translation file is to add an additional `context-group` to each `trans-unit` where we could add e.g. the corresponding question, the answers and so on.

Reply:
1. With the manual approach, we are currently using nobody can reuse the ids, because nobody will be able to match them with one another. 
2. Further, the question id should be unique and no question with the same text should ever exist. By reusing the ids of the current element we can easily ensure that, while with MD5 Hashes we would have to incorporate another check. 
3. After committing a questionnaire into the GitHub repo the strings in source version should be replaced with the unique ID from the translation file.

Reply:
1): I dont know. The translation can be reused as soon as the hash matches which means that it can be reused as soon as the questions are identical. I see problems when someone changes the text of a question but forgets to change the corresponding translation.
2): What do you want to check? I mean the MD5 ensures that the translation always fits the current text. Since the translation ID is not reused anywhere, we do not have to provide checks at all.
3): Why? I mean maintaining is much easier when you can read the question itself and not a ID. And with the current approach (but also with the other approach) there is no need to replace the strings.

In addition, I believe that the assumption in 2 is not correct. I guess, that people will start to duplicate questions to ensure their dialogue flow (see FIHR discussion) or because they formulate follow up questions (e.g. smth like _When was it?_, bad design, but with our standard valid).  A third possiblity are answer options like _I dont know_ and so on which are also not unique. So I dont think that we will have unique IDs together with unique text, so I wont mix up question/answer IDs and their text.

Reply:
Consider the following example: 
```jsonc
// Question:
  {
      "id": "q_contact",
      "text": "Gab es Kontakt zu bestätigten Fällen?", // Hash: g234jl345lk4j45ljlk324jjlkj345lk
      "type": "boolean"
  },
  {
      "id": "q_contact",
      "text": "Gab es Kontakt ins Ausland?", // Hash: abfd245m234l234n2435kjh5345
      "type": "boolean"
  },
  {
      "id": "q_contact",
      "text": "Gab es Kontakt ins Ausland?", // trans-id: q_contact-text
      "type": "boolean"
  }

```
If we now change the text of the question but leave the question id as it is, we have two different questions with the same id and we would have to develop another script to check for duplicates. 

If we use the id of the question as translation id error handling is much easier as we simply disallow a second usage. 

> 3): Why? I mean maintaining is much easier when you can read the question itself and not a ID. And with the current approach (but also with the other approach) there is no need to replace the strings.

You have a point. But consider changing an English sentence (which basically is our id, as we are generating an MD5 from it) in all questionnaires. Instead of changing it once in the translation file, we would have to change it in each questionnaire. Which kind of defeats the purpose of a translation file. 

For the rest you are right, that's basically the same for both solutions.

Reply:
I am not sure if I get your example right, but is the idea now to disallow duplicate question ids by reusing the same translation so that the question text is the same in the end? (Correct me if I am wrong). I think we need a check for duplicate question IDs anyway (actually we need much more advanced checks, e.g. if all question IDs showing up in the variables really exist in the questionnaire). So I would suggest to keep translations out of this.

Do you plan to reuse questions across questionnaires by reusing also their ID? As far as I understood, reusing questions was just about copying the question from one questionnaire to another, but keep them independent afterwards. Then it makes sense to replace the strings with IDs and to use readable IDs, but maybe we should discuss if this is a good idea actually. Example: Doctor A creates a question about symptoms, Doctor B reuses the question. Then Doctor B decides to add another possible symptom. This will show up also for Doctor A although they did not program this in first place.  In worst case, this could actually break their JSON Logic for the results! So I dont think this is a good idea, but maybe I got you wrong here as well.

What we could consider is to share translation files between all questionnaires, but again - this would require MD5 :D Then people can set their desired question IDs, they can add/remove options and so, but as long as they use the same string (with the same hash) they can reuse translations safely.

Reply:
So maybe we should clarify, if question IDs are unique across questionnaires. I think we should not enforce this (because it would require super specific ids and not something like `q1_contact`. What do you think? Do you plan to share IDs between questionnaires?

Reply:
> So maybe we should clarify, if question IDs are unique across questionnaires. I think we should not enforce this (because it would require super specific ids and not something like `q1_contact`. What do you think? Do you plan to share IDs between questionnaires?

That's what I thought was the base we are all building on? Most of the questionnaires will share the same logic and if the responses generated should be comparable by any standard the question ids have to be the same. Otherwise, you will come into mapping hell.

Reply:
I would go on with your implementation just that I will only use the initial translation for id generation from then on the original resource will have a string like
```json
"text": "This is the original English sentence.@@MD5Hashöldfks53kj345lkjl4356lö"
```

Reply:
Just to be sure: this additional hash/guid/uuid will not be exposed to external apps that consume the API, but it is just an internal representation of the data, correct?

Reply:
it will not be exposed for endpoint asking for a specific language, for everything else it will be exposed. Have a look at the test directory for a example.